From 7891892667e172d4c586e2d66c6a2a457ac37ba8 Mon Sep 17 00:00:00 2001
From: fli <fupan.li@windriver.com>
Date: Thu, 20 Jul 2017 22:29:54 -0700
Subject: [PATCH] python-smartpm:Don't ignore error if RPM transaction fails
 without problems

SmartPM could misinterpret RPM transaction error as success,
if ts.run() (RPM Python API) returns an empty problems list.

This could happen for example if the RPM pkg's postins scripts
executed failed, it's ts.run() will return an empty problem list.

Signed-off-by: fli <fupan.li@windriver.com>
---
 smart/backends/rpm/pm.py | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/smart/backends/rpm/pm.py b/smart/backends/rpm/pm.py
index cae13e8..79592cb 100644
--- a/smart/backends/rpm/pm.py
+++ b/smart/backends/rpm/pm.py
@@ -210,7 +210,7 @@ class RPMPackageManager(PackageManager):
         force = sysconf.get("rpm-force", False)
         if not force:
             probs = ts.check()
-            if probs:
+            if (probs is not None) and (len(probs) != 0):
                 problines = []
                 for prob in probs:
                     name1 = "%s-%s-%s" % prob[0]
@@ -259,7 +259,7 @@ class RPMPackageManager(PackageManager):
         finally:
             del getTS.ts
             cb.grabOutput(False)
-            if probs and sysconf.has("attempt-install", soft=True):
+            if (probs is not None) and sysconf.has("attempt-install", soft=True):
                 def remove_conflict(pkgNEVR):
                     for key in changeset.keys():
                         if pkgNEVR == str(key):
@@ -290,7 +290,7 @@ class RPMPackageManager(PackageManager):
                         retry = 0
 
             prog.setDone()
-            if probs and (not retry):
+            if (probs is not None) and (not retry):
                 raise Error, "\n".join([x[0] for x in probs])
             prog.stop()
             if retry and len(changeset):
-- 
2.11.0

